### ISSUE TITLE
[FireMonkey] Request YT' JS script ad blocker

### ISSUE THREAD

**Issue by EstherMoellman** [erosman/support#247]
Hi @erosman ! How are you?

For the first time, today I found a JS script that can block all YT' video ads.
I tested every single YT' JS script ads blocker, they do block YT' "skippable" ads, but they can't block YT' "non-skippable" ads.
This JS script can block both types of YT' ads: https://greasyfork.org/en/scripts/32626-disable-youtube-video-ads

It works perfectly fine with Firefox Nightly + TamperMonkey.
However, unfortunately I can't make it work with FireMonkey (YT' pages are appearing almost empty).

Please, can you help me to tweak this JS script in order to make it work with FireMonkey?
Or please, can you adapt FireMonkey in order to work with this JS script?
If by chance you can give me a hand... it'll be great! : )

Currently I'm using an add-on to block YT' non-skippable ads.
But if you help me, and if this JS script can work with FireMonkey, then I'll be able to remove the add-on (and removing add-ons is important for me, because I'm a minimalist girl (LOL), if you remember, I always try to keep the less number of add-ons as possible, all my Nightly' settings are focused on browser performance). And FireMonkey here has been a savior for me, because it helps me to replace lot of add-ons.

Thanks in advance and big hug! : )

**Comment by erosman** [comment]
Try this and see what happens:

```
// @run-at      document-end
```

**Comment by EstherMoellman** [comment]
@erosman , thanks for your replay!

Before opening the issue here, I did a lot of tests, including the "// @run-at      document-end"... but nothing worked.
Long time ago you explained me the "document-start" issue with FireMonkey. But honestly, this issue is very random, because in my FireMonkey I have several JS scripts working perfectly with "document-start". Anyway, I always remember the "document-start" issue, and when I have difficulties testing a JS script on FireMonkey, the first thing to remove/change is the "document-start".
In short, "document-end" in this JS script allows the YT' page to load, but the ad blocker function stops working.

Hope you can find a solution : )
Thks a lot!

**Comment by erosman** [comment]
I recall this script from before. As I mentioned before, script developer has more intimate knowledge about the script and can debug better.

When running the script it results in `Uncaught ReferenceError: ipse is not defined` which relates to line 71

```js
if (!ipse && (a = document.querySelector('ytd-watch-flexy')) && (a = a.constructor.prototype) && a.isPlaShelfEnabled_) {
```

Check with developer to see if that helps.

**Comment by EstherMoellman** [comment]
@erosman , what a memory you have... good for you! LOL

Yeap, this script is from same author, but during the past months this script had lot of updates, upgrades and corrections. And only today, its last version, finally blocked both type of YT' ads (skippable and non-skippable).

And again, before opening the issue here, I always try first to contact the Dev.
And the Dev is a nice person, he uses to replay my comments, but he doesn't seem interested in specific adaptations (for example with FireMonkey). This always happened in the past, I contacted him at Reddit and Greasyfork.org, he is very polite and responsive, but also he is very clear, not interested in specific add-ons etc.

Today early I already wrote him a comment, I'll add your info (THANK YOU!), but if by chance you find something helpful, please share with me, it'll help me to remove one add-on, keeping my browser better and better in terms of performance and minimalism.

Feel free to close the issue whenever you want.

Big hug again!

**Comment by erosman** [comment]
The script that I have written, all work the same way on FM/GM/VM/TM. There is no need to write a specific script for each script-manager.

In fact, my shared scripts work even on older GM3.

One possibility is that since other script manager inject the script manually (not using the dedicated API), the script gets injected a bit later than in FM.

You can ask the dev to try `'use strict';` which can show some of the common errors.

**Comment by EstherMoellman** [comment]
... thanks a lot @erosman for your always useful help!

I added all your explanations in my comment to the Dev of the JS script.
As I said, I'm sure he'll replay me, but sadly, I'm not sure he'll be interested in FM.
Even if your explanations are not related to FM, the script works fine with TamperMonkey, and Dev seems not interested in other compatibilities. Hope I'm wrong!

Many times I saw this behavior where a Dev focuses just on a specific browser or Monkey add-on/extension etc, and if it works for there, then the Dev stops there, no interest in different browsers or different Monkey add-on/extension incompatibilities.
Worse, many Dev/Users judge FM based on popularity (number of users), and wrongly they conclude it's not worth to them to invest solving compatibilities or other issues.
The same happens with browsers, if a JS script works for GoogleChrome, then most of the Devs stop there and are satisfied.

But I don't blame them at all. I understand, Devs are working for free, and makes sense to focus on compatibilities with the most used browsers/monkey extensions.
My only concern is to see Firefox or great add-ons/extensions disappearing just because software ecosystem with time is more and more smaller.

**Comment by erosman** [comment]
Does the script work in GM?

**Comment by EstherMoellman** [comment]
> 
> 
> Does the script work in GM?

Nope.
Similar behavior than FM.
With GM, YT' pages load almost blank. If I change to "document-end", YT' page loads but ad blocker doesn't work.

**Comment by erosman** [comment]
Therefore, it is not specific to FM. ;)

**Comment by EstherMoellman** [comment]
TM (TamperMonkey): Works perfectly
VM (ViolentMonkey): Works perfectly
GM (GreaseMonkey) + FM (FireMonkey): Not working

PS: Sometimes the problem is not related just to monkey-add-on incompatibilities, but to the combination of monkey-add-on + Firefox. I mean, in tests with other scripts, for example, in the past I saw scripts working with monkey-extension + GoogleChrome, while with same monkey-add-on + Firefox the script is not working.

However, IMHO, it doesn't matter the nature of the incompatibility (monkey-add-on or browser). The real problem is that most of the Dev seem to be satisfied if their script works a specific monkey-extension + specific browser.

**Comment by erosman** [comment]
In fact, it is proper JavaScript. Variables have to be declared.

Since 2015, the proper method of writing JavaScript is to use `'use strict';`
Using `'use strict';` throws errors & halts the script if there are issues with the code (like undeclared variables).
This leads to a cleaner and error-free code.

Some JavaScript writer avoid `'use strict';` in order to avoid updating their code.

Firefox userScript API injects the code as `'use strict';` (in strict mode)
That way, such errors breaks the code.

GM might be using `'use strict';` as well.

This is not a compatibly issue. It is simply good (and modern) code writing.

When `'use strict';`  is not used, the error still occurs, but is doesn't halt the script.

There are also scope issues (blocked vs global) with such behaviour (as well as using global `var`) which affects the performance of the script in the browser.

**Comment by EstherMoellman** [comment]
@erosman , I loved your explanation! Thanks!
This script does the job, for the first time at my YT all ads are blocked, including skippable, unskippable and CSS ads. It's the first time I see this kind of complete YT' ads blocker, all-in-one js script.
As your final touch in this script, do you want to delete, change or replace something else in order to make it even better, or improve performance etc?

**Comment by erosman** [comment]
> As your final touch in this script, do you want to delete, change or replace something else in order to make it even better, or improve performance etc?

TBH, it is better for the the developer to do it. I only added the line to make it work. That is all. ;)
